File: V0.1/apps/knowledge-radar-agent/src/knowledge_radar/integrations/llm_client.py
"""
LLM Client - 大语言模型客户端

封装与 LLM 服务的通信
"""

from typing import Dict, Any, Optional, Type, List
from pydantic import BaseModel, ValidationError
import json
import asyncio


class LLMClient:
    """
    LLM 客户端
    
    支持 OpenAI-compatible chat completions API
    提供 generate_text 和 generate_json 方法
    """
    
    def __init__(
        self,
        api_base: str = "https://api.openai.com/v1",
        api_key: Optional[str] = None,
        model: str = "gpt-4",
        max_retries: int = 3,
    ):
        self.api_base = api_base
        self.api_key = api_key
        self.model = model
        self.max_retries = max_retries
    
    async def generate_text(
        self,
        prompt: str,
        system_prompt: Optional[str] = None,
        temperature: float = 0.7,
        max_tokens: int = 4096,
    ) -> str:
        """
        生成文本
        
        Args:
            prompt: 用户提示
            system_prompt: 系统提示
            temperature: 温度参数
            max_tokens: 最大 token 数
        
        Returns:
            生成的文本
        """
        messages = []
        
        if system_prompt:
            messages.append({"role": "system", "content": system_prompt})
        
        messages.append({"role": "user", "content": prompt})
        
        response = await self._call_api(messages, temperature, max_tokens)
        
        return response.get("choices", [{}])[0].get("message", {}).get("content", "")
    
    async def generate_json(
        self,
        prompt: str,
        schema: Type[BaseModel],
        system_prompt: Optional[str] = None,
        temperature: float = 0.1,
        max_tokens: int = 4096,
    ) -> BaseModel:
        """
        生成 JSON 并校验
        
        Args:
            prompt: 用户提示
            schema: Pydantic schema
            system_prompt: 系统提示
            temperature: 温度参数
            max_tokens: 最大 token 数
        
        Returns:
            校验后的 Pydantic 对象
        
        Raises:
            ValueError: 如果 JSON 解析或校验失败
        """
        system_message = f"""
你必须输出严格符合以下 JSON Schema 的数据：
{schema.model_json_schema()}

输出格式必须是纯 JSON，不要包含任何其他文本。
"""
        
        if system_prompt:
            system_message = f"{system_prompt}\n\n{system_message}"
        
        messages = [
            {"role": "system", "content": system_message},
            {"role": "user", "content": prompt},
        ]
        
        response = await self._call_api(messages, temperature, max_tokens)
        content = response.get("choices", [{}])[0].get("message", {}).get("content", "")
        
        try:
            data = json.loads(content)
            return schema(**data)
        except (json.JSONDecodeError, ValidationError) as e:
            raise ValueError(f"JSON 解析或校验失败: {str(e)}")
    
    async def _call_api(
        self,
        messages: List[Dict[str, str]],
        temperature: float,
        max_tokens: int,
    ) -> Dict[str, Any]:
        """
        调用 LLM API
        
        Args:
            messages: 消息列表
            temperature: 温度参数
            max_tokens: 最大 token 数
        
        Returns:
            API 响应
        """
        for attempt in range(self.max_retries):
            try:
                # 模拟 API 调用
                # 实际实现时需要使用 httpx 或 requests 调用真实 API
                return self._mock_response(messages)
            except Exception as e:
                if attempt < self.max_retries - 1:
                    await asyncio.sleep(2 ** attempt)
                else:
                    raise e
    
    def _mock_response(self, messages: List[Dict[str, str]]) -> Dict[str, Any]:
        """
        模拟 API 响应
        
        Args:
            messages: 消息列表
        
        Returns:
            模拟的 API 响应
        """
        user_content = ""
        for msg in messages:
            if msg["role"] == "user":
                user_content = msg["content"]
                break
        
        # 简单模拟响应
        return {
            "choices": [
                {
                    "message": {
                        "content": f"这是对请求的响应: {user_content[:50]}...",
                    }
                }
            ]
        }

File: backend-server/llm-client.js
/**
 * Knowledge Radar — LLM Client
 *
 * Entity extraction, knowledge summarization, and relationship inference.
 * Supports OpenAI-compatible APIs (OpenAI, DeepSeek, Ark, etc.).
 */

import fetch from 'node-fetch';

export class LLMClient {
  constructor(config = {}) {
    this.baseUrl = config.baseUrl || process.env.LLM_BASE_URL || 'https://api.openai.com/v1';
    this.apiKey = config.apiKey || process.env.LLM_API_KEY || '';
    this.model = config.model || process.env.LLM_MODEL || 'gpt-4o-mini';
    this._available = !!this.apiKey;
  }

  get available() { return this._available; }

  async chat(messages, options = {}) {
    if (!this._available) return null;
    
    const body = {
      model: this.model,
      messages,
      temperature: options.temperature ?? 0.1,
      max_tokens: options.maxTokens ?? 4096,
      response_format: options.jsonMode ? { type: 'json_object' } : undefined,
    };

    try {
      const res = await fetch(`${this.baseUrl}/chat/completions`, {
        method: 'POST',
        headers: {
          'Authorization': `Bearer ${this.apiKey}`,
          'Content-Type': 'application/json',
        },
        body: JSON.stringify(body),
      });
      if (!res.ok) {
        const text = await res.text().catch(() => '');
        console.warn(`[LLM] API error (${res.status}): ${text.slice(0, 200)}`);
        return null;
      }
      const data = await res.json();
      return data.choices?.[0]?.message?.content || null;
    } catch (e) {
      console.warn(`[LLM] Request failed:`, e.message);
      return null;
    }
  }

  // ── Entity Extraction from Message ──────────────────────────────────────

  async extractEntitiesFromMessage(message) {
    const prompt = `Extract all entities and their relationships from this IM message.

Message: "${message.content}"

Entity types to look for: person, project, task, concept, decision, document, meeting, risk, system

Respond with JSON only:
{
  "entities": [{ "name": "...", "type": "...", "aliases": ["..."] }],
  "relations": [{ "source": "...", "target": "...", "type": "...", "description": "..." }],
  "knowledge": [{ "type": "decision|action_item|risk|update|faq|info", "title": "...", "summary": "..." }]
}`;

    const result = await this.chat([
      { role: 'system', content: 'You are a knowledge extraction engine. Output only valid JSON.' },
      { role: 'user', content: prompt },
    ], { jsonMode: true });

    if (!result) return null;
    try {
      return JSON.parse(result);
    } catch {
      console.warn('[LLM] Failed to parse entity extraction result');
      return null;
    }
  }

  // ── Knowledge Summarization ─────────────────────────────────────────────

  async summarizeKnowledge(items, context) {
    const prompt = `Summarize the following knowledge items into a coherent briefing.

Context: ${context}

Items:
${items.map((i, idx) => `${idx + 1}. [${i.type}] ${i.title}: ${i.summary}`).join('\n')}

Respond with JSON:
{
  "summary": "brief summary paragraph",
  "keyPoints": ["point 1", "point 2"],
  "suggestedActions": ["action 1"]
}`;

    const result = await this.chat([
      { role: 'system', content: 'You are a knowledge management assistant. Output only valid JSON.' },
      { role: 'user', content: prompt },
    ], { jsonMode: true });

    if (!result) return null;
    try {
      return JSON.parse(result);
    } catch {
      return null;
    }
  }

  // ── Meeting Briefing Generation ─────────────────────────────────────────

  async generateMeetingBriefing(meeting, previousMeeting, recentKnowledge) {
    const prompt = `Generate a pre-meeting briefing based on the following data.

Current meeting: ${meeting.title} at ${meeting.startTime}
Description: ${meeting.description || ''}
Participants: ${(meeting.participants || []).map(p => p.name || p).join(', ')}

${previousMeeting ? `Previous meeting decisions:\n${previousMeeting.decisions?.map(d => `- ${d}`).join('\n') || 'None'}\n\nPrevious action items:\n${previousMeeting.actionItems?.map(a => `- ${a.text} (${a.owner}, ${a.status})`).join('\n') || 'None'}` : 'No previous meeting data.'}

${recentKnowledge.length ? `Recent related knowledge:\n${recentKnowledge.map(k => `- [${k.knowledge_type}] ${k.title}: ${k.summary}`).join('\n')}` : ''}

Respond with JSON:
{
  "summary": "Briefing summary in markdown",
  "lastMeetingDecisions": ["decision 1"],
  "pendingActionItems": [{"text": "...", "owner": "..."}],
  "relatedUpdates": [{"title": "...", "detail": "..."}],
  "suggestedTopics": ["discussion topic"]
}`;

    const result = await this.chat([
      { role: 'system', content: 'You are a meeting preparation assistant. Output only valid JSON.' },
      { role: 'user', content: prompt },
    ], { jsonMode: true });

    if (!result) return null;
    try {
      return JSON.parse(result);
    } catch {
      return null;
    }
  }

  // ── Weekly Digest Generation ────────────────────────────────────────────

  async generateWeeklyDigest(knowledgeItems, entities) {
    const prompt = `Generate a weekly knowledge digest from the following data.

Knowledge items this week:
${knowledgeItems.map((k, i) => `${i + 1}. [${k.knowledge_type}] ${k.title}: ${k.summary}`).join('\n')}

Active entities:
${entities.map(e => `- ${e.name} (${e.entity_type})`).join('\n')}

Respond with JSON:
{
  "title": "Weekly Knowledge Digest",
  "sections": [
    {"heading": "...", "items": ["item1", "item2"]}
  ],
  "summary": "overall summary"
}`;

    const result = await this.chat([
      { role: 'system', content: 'You are a knowledge management assistant. Output only valid JSON.' },
      { role: 'user', content: prompt },
    ], { jsonMode: true });

    if (!result) return null;
    try {
      return JSON.parse(result);
    } catch {
      return null;
    }
  }
}
